martin-mbtiles/src/mbtiles.rs

```
@@ -378,32 +388,102 @@ impl Mbtiles {
+
+        }.to_string();
+
+        let rusqlite_conn = RusqliteConnection::open(Path::new(&self.filepath()))?;
+        register_md5_function(&rusqlite_conn)?;
+        rusqlite_conn.execute(
+            format!("INSERT OR REPLACE INTO metadata(name, value) {select_from}").as_str(),
+            [],
+        )?;
+        Ok(())
+    }
+
+    pub async fn validate_mbtiles<T>(
+        &self,
+        integrity_check: IntegrityCheck,
+        conn: &mut T,
+    ) -> MbtResult<()>
+    where
+        for<'e> &'e mut T: SqliteExecutor<'e>,
+    {
+        // SQLite Integrity check
+        if "ok"
+            != match integrity_check {
+                IntegrityCheck::Full => query("PRAGMA integrity_check;")
+                    .fetch_one(&mut *conn)
+                    .await?
+                    .get::<String, _>(0),
+                IntegrityCheck::Quick => query("PRAGMA integrity_check;")
+                    .fetch_one(&mut *conn)
+                    .await?
+                    .get::<String, _>(0),
+                IntegrityCheck::Off => "ok".to_string(),
             }
+        {
+            return Err(FailedIntegrityCheck(self.filepath().to_string()));
+        }
+
+        let mbttype = self.detect_type(&mut *conn).await?;
+
+        if mbttype == MbtType::Flat {
+            println!(
+                "No hash column in flat table format, skipping hash-based validation steps..."
+            );
+            return Ok(());
         }
-        .to_string();
 
         let rusqlite_conn = RusqliteConnection::open(Path::new(self.filepath()))?;
         register_md5_function(&rusqlite_conn)?;
 
+        // Global hash check
+        if self
+            .get_metadata_value(&mut *conn, "global_hash")
+            .await?
+            .is_none()
+        {
+            println!(
```

if the global hash is missing, its an error. We do not want to rely on silent failure. If used programmatically, noone will ever see these print statements, but they will catch non-zero program exit codes

---

If we fail everytime the `global_hash` value is missing, we can never generate the global hash value for the first time because we will fail before we get to that step. I agree it is good to just fail here but I'm not sure of a cleaner way to actually generate the global hash for the first time.

---

I will make the global hash generation happen at the copying level